api/client/service/ps.go

```diff
@@ -44,13 +48,34 @@ func runPS(dockerCli *client.DockerCli, opts psOptions) error {
 	client := dockerCli.Client()
 	ctx := context.Background()
 
-	service, _, err := client.ServiceInspectWithRaw(ctx, opts.serviceID)
-	if err != nil {
-		return err
+	filter := opts.filter.Value()
+	filterTypes := []string{"id", "name"}
+
+	unknownServices := []string{}
+
+service_search:
+	for _, service := range opts.services {
+		for _, filterType := range filterTypes {
+			serviceFilter := filters.NewArgs()
+			serviceFilter.Add(filterType, service)
+			if serviceList, err := client.ServiceList(ctx, types.ServiceListOptions{Filter: serviceFilter}); err == nil && len(serviceList) != 0 {
+				for _, serviceEntry := range serviceList {
+					filter.Add("service", serviceEntry.ID)
+				}
+				continue service_search
```

Comment:
@yongtang I can read the code. I mean "What is the purpose of this loop?". It's fairly obtuse.